### PR TITLE
Backport sdf6: Move recursiveSameTypeUniqueNames from ign.cc to parser.cc

### PR DESCRIPTION
**Pull request by j-rivero**
# 🎉 New feature

## Summary
Backport of sdf9 PR 606 https://osrf-migration.github.io/sdformat-gh-pages/#!/osrf/sdformat/pull-requests/606
Corresponding commit: https://github.com/osrf/sdformat/commit/2d1bc5627542f4072b1c2c5628074e8e1120a93d

## Test it
The backport include its own test

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Code check passed (In source directory, run `sh tools/code_check.sh`)
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Comment by j-rivero** [comment]
something weird happened to github actions. Close and reopen to trigger them.

File: Changelog.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ### SDFormat 6.X.X (20XX-XX-XX)
 
+1. Move recursiveSameTypeUniqueNames from ign.cc to parser.cc and make public.
+    * [Pull request ]()
```
Comment (iche033) [inline Changelog.md:6]: reminder to update this before release

**Review by iche033 — APPROVED** [review]
look good to me

File: Changelog.md (hunk 1 above)
Comment (j-rivero) [reply Changelog.md]: I was about to forget about it. Thanks Ian, done.

**Review by jennuine — APPROVED** [review]

**Pull request merged**